core-plugins/src/test/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSourceTest.java

```diff
@@ -0,0 +1,425 @@
+          }
+        }
+      }
+    }
+    return processedFileList;
+  }
+
+  @Test
+  /**
+   * This test validate following
+   * 1. Read multiple files from folder
+   * 2. Delete files once processed
+   * 3. Filter Pre-Processed file
+   * 4. Delete 40 days old record from file tracking table.
+   */
+  public void testXMLReaderWithNoXMLPreProcessingRequired() throws Exception {
+    String processedFileTable = "XMLTrackingTableNoXMLPreProcessingRequired";
+    Map<String, String> sourceProperties = new ImmutableMap.Builder<String, String>()
+      .put(Constants.Reference.REFERENCE_NAME, "XMLReaderNoXMLPreProcessingRequiredTest")
+      .put("path", sourceFolderUri)
+      .put("nodePath", "/catalog/book/price")
+      .put("targetFolder", targetFolderUri)
+      .put("reprocessingRequired", "No")
+      .put("tableName", processedFileTable)
+      .put("actionAfterProcess", "Delete")
+      .put("tableExpiryPeriod", "40")
+      .build();
+
+    ETLStage source = new ETLStage("XMLReader", new ETLPlugin("XMLReader", BatchSource.PLUGIN_TYPE,
+                                                              sourceProperties, null));
+
+    String outputDatasetName = "output-batchsink-test-no-preprocessing-required";
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputDatasetName));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder("* * * * *")
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    AppRequest<ETLBatchConfig> appRequest = new AppRequest<>(ETLBATCH_ARTIFACT, etlConfig);
+    ApplicationId appId = NamespaceId.DEFAULT.app("XMLReaderNoXMLPreProcessingRequiredTest");
+    ApplicationManager appManager = deployApplication(appId.toId(), appRequest);
+
+    Date preProcessedDate = new Date();
+    createPreProcessedRecord(processedFileTable, preProcessedDate);
+    createExpiredRecord(processedFileTable);
+
+    MapReduceManager mrManager = appManager.getMapReduceManager(ETLMapReduce.NAME);
+    mrManager.start();
+    mrManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    //assert for number of files processed
+    List<String> processedFileList = getProcessedFileList(processedFileTable, preProcessedDate);
+    Assert.assertEquals(1, processedFileList.size());
+
+    //assert for number of record derived.
+    DataSetManager<Table> outputManager = getDataset(outputDatasetName);
+    List<StructuredRecord> output = MockSink.readOutput(outputManager);
+    Assert.assertEquals(3, output.size());
+    //source folder left with one pre-processed file
+    File[] sourceFiles = sourceFolder.listFiles();
+    Assert.assertEquals(1, sourceFiles.length);
+  }
+
+  @Test
+  /**
+   * This test validate following
+   * 1. Read multiple files from folder
+   * 2. PreProcessing of files
+   */
+  public void testXMLReaderWithXMLPreProcessingRequired() throws Exception {
+    String processedFileTable = "XMLTrackingTableXMLPreProcessingRequired";
```

Comment:
Please check if you can refactor some common code in this and the following tests to a method to reduce duplication.